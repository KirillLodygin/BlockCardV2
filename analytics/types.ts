import { NOTIFICATION_TYPE } from '@vtb/ui-kit/adaptive/molecules/notification';

import { LOCK_OPERATION } from '../types';
import {
  CREDIT_ACTION_PERMANENT_BLOCK_STRING,
  CREDIT_ACTION_TEMPORARY_BLOCK_STRING,
  CREDIT_CARD_ACTION_MAP,
  CREDIT_PRODUCT,
  CREDIT_STREAM,
  DEBIT_ACTION_PERMANENT_BLOCK_STRING,
  DEBIT_ACTION_TEMPORARY_BLOCK_STRING,
  DEBIT_CARD_ACTION_MAP,
  DEBIT_PRODUCT,
  DEBIT_STREAM,
} from './constants';

export type AnalyticAction = {
  text: string;
  target: string;
};

// для аналитики
export enum CARD_ACTION_TYPE {
  UNFREEZE = 'unfreeze',
  BLOCK = 'block',
}

export type AnalyticsMapType = {
  [x in string]: {
    [x in string]: {
      text: string;
      target: string;
    };
  };
};

export type AnalyticsEntityType = {
  stream: typeof CREDIT_STREAM | typeof DEBIT_STREAM;
  product: typeof CREDIT_PRODUCT | typeof DEBIT_PRODUCT;
  getActionEntityString: (
    blockType: LOCK_OPERATION,
  ) =>
    | typeof DEBIT_ACTION_TEMPORARY_BLOCK_STRING
    | typeof DEBIT_ACTION_PERMANENT_BLOCK_STRING
    | typeof CREDIT_ACTION_TEMPORARY_BLOCK_STRING
    | typeof CREDIT_ACTION_PERMANENT_BLOCK_STRING;
  analyticsMap: AnalyticsMapType;
  type: typeof CARD_ACTION_TYPE.BLOCK | typeof CARD_ACTION_TYPE.UNFREEZE;
};

enum BLOCK_DEBIT_CARD_SCREEN {
  CHECK = 'CHECK',
  RESULT = 'RESULT',
  START = 'START',
}

export type NotificationType = typeof NOTIFICATION_TYPE.info | typeof NOTIFICATION_TYPE.fail;

export type AnalyticsOptionsType = {
  blockType: LOCK_OPERATION;
  cardName: string;
  screenData: BLOCK_DEBIT_CARD_SCREEN;
  notificationType?: NotificationType;
  reason?: string;
};

export type DebitAnalyticsMapType = typeof DEBIT_CARD_ACTION_MAP;
export type CreditAnalyticsMapType = typeof CREDIT_CARD_ACTION_MAP;

export type NotificationKeys = keyof typeof DEBIT_CARD_ACTION_MAP;
