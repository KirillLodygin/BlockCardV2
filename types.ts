import { DropdownValue } from '@vtb/ui-kit/adaptive/molecules/dropdown/_baseDropdown/typesDropdown';

import { ServiceListMsa } from '@modules/CreditCard';
import { EntityByType } from '@modules/DebitAndCreditCards';

export type BlockCardReqeustDataType = {
  secureCode: string;
  transactionId: string;
};

export enum SCENARIO_ASYNC_EVENT_BLOCK_CARD {
  AUTHORIZE = 'card-lock-authorize',
  AUTHORIZE_UNFREEZE = 'card-unfreeze-authorize',
  COMMIT = 'card-lock-commit',
  COMMIT_UNFREEZE = 'card-unfreeze-commit',
  FREEZE = 'card-freeze',
  UNFREEZE = 'card-unfreeze',
}

export type AsyncEventBlockDebitCardResponse = {
  body: AuthorizeResponseType | BlockCardResponseType;
  status: number;
  type: SCENARIO_ASYNC_EVENT_BLOCK_CARD;
};

export type BlockCardResponseType = {
  maskedCardNumber: string;
};

export type AuthorizeResponseType = {
  expirationTime: number;
  phone: string;
  transactionId: string;
};

export type RowTableData = {
  value: string | DropdownValue | null;
  title: string;
};

export enum BLOCK_DEBIT_CARD_SCREEN {
  CHECK = 'CHECK',
  RESULT = 'RESULT',
  START = 'START',
}

export enum BLOCK_TABLE_DATA_TITLE {
  BLOCK_REASON = 'blockReason',
  CARD_TITLE = 'blockCardTitle',
  OPERATION_TYPE = 'operationType',
  SHROT_CRAD_NEMBER = 'shortNumber',
}

export enum LOCK_OPERATION {
  PERMANENT = 'permanent',
  TEMPORARY = 'temporary',
  UNFREEZE = 'unfreeze',
}

export enum BUTTONS_LOCK {
  BACK = 'back',
  CONTINUE = 'continue',
  RESEND = 'resend',
  SECONDARY = 'secondary',
}

export enum NOTIFICATION_TYPE {
  AUTH_ERROR = 'authError',
  ERROR = 'error',
  FAIL = 'fail',
  INFO = 'info',
  SUCCESS = 'success',
}

export enum CARD_TYPE {
  DIGITAL = 'digital',
  PLASTIC = 'plastic',
}

export enum LOCK_REASON {
  LOSS = 'loss',
  THEFT = 'theft',
  FRAUD = 'fraud',
  STUFFY = 'stuffy',
  OTHER = 'other',
}

export type ButtonsLock = {
  isFail?: boolean;
  title: string;
  type: BUTTONS_LOCK;
};

export type NotificationLockResult = {
  description: string;
  operations: LOCK_OPERATION[];
  title?: string;
  type: NOTIFICATION_TYPE;
};

export type LockTypeSelection = {
  buttons: ButtonsLock[];
  buttonsRadio: {
    id: LOCK_OPERATION;
    text: string;
  }[];
  dropdown: {
    elements: {
      id: LOCK_REASON;
      cardTypes: CARD_TYPE[];
      text: string;
    }[];
    label: string;
  };
  inputLabel: string;
  notifications: NotificationLockResult[];
  title: string;
};

export type LockConfirmationScreen = {
  buttons: ButtonsLock[];
  inputErrorText: string;
  inputLabel: string;
  operationName: string;
  phoneInfo: string;
  timerInfo: string;
  table: {
    title: string;
    type: BLOCK_TABLE_DATA_TITLE;
  }[];
  title: string;
};

export type LockResultScreen = {
  buttons: ButtonsLock[];
  notifications: NotificationLockResult[];
};

export type ContentCardBlock = {
  lockTypeSelectionScreen: LockTypeSelection;
  lockConfirmationScreen: LockConfirmationScreen;
  lockResultScreen: LockResultScreen;
};

export type ContentCreditCardBlock = {
  BlockCardHeader?: string;
  BlockCardTemp?: string;
  BlockCardConst?: string;
  BlockCardButton?: string;
  BlockCardDescription?: string;
  TempBlockCardDescription?: string;
  BlockCardConfirmButton?: string;
  BlockCardConfirmHeader?: string;
  BlockCardConfirmTitle?: string;
  BlockCardConfirmNumber?: string;
  BlockCardConfirmType?: string;
  BlockCardConfirmReason?: string;
  SuccessBlockCardHeader?: string;
  SuccessBlockCardDescription?: string;
  SuccessButton?: string;
  ErrorBlockCardHeader?: string;
  ErrorBlockCardDescription?: string;
  ErrorButton?: string;
  NetworkErrorHeader?: string;
  NetworkErrorDescription?: string;
  NetworkErrorAgainButton?: string;
  TempBlockCardHeader?: string;
  SuccessTempBlockCardDescription?: string;
  UnblockCardHeader?: string;
  UnblockCardTitle?: string;
  UnblockCardNumber?: string;
  UnblockCardButton?: string;
  SuccessUnblockCardHeader?: string;
  SuccessUnblockCardDescription?: string;
  ErrorUnblockCardHeader?: string;
  ErrorUnblockCardDescription?: string;
  BlockUserHeader?: string;
  EBlockUserDescription?: string;
  CauseBlockArray?: {
    id?: 'first' | 'second' | 'third' | 'fourth';
    title?: string;
    enabled?: boolean;
    reasonFor?: 'PLASTIC' | 'ALL';
  }[];
};

export type CreditCard = EntityByType<'CreditCard'> & { serviceList?: ServiceListMsa };
