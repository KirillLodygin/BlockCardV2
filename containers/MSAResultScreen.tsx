import React from 'react';
import { Box, Flex } from 'reflexbox/styled-components';

import { getDataTestId } from '@vtb/ui-kit/_utils/getDataTestId';
import { ButtonContained, ButtonOutlined } from '@vtb/ui-kit/adaptive/atoms/button';
import { NOTIFICATION_TYPE } from '@vtb/ui-kit/adaptive/molecules/notification';
import { FIELD_WIDTH, INDENT } from '@vtb/ui-kit/tokens/sizes';

import { useBreakpointState } from '@common/hooks/useBreakpoint';

import { AdaptiveOperationStatus } from '../components/AdaptiveOperationStatus';
import { ResultScreenSkeleton } from '../components/ResultScreenSkeleton';
import { BUTTONS_LOCK } from '../types';

type NotificationType = typeof NOTIFICATION_TYPE.info | typeof NOTIFICATION_TYPE.fail;

type Props = {
  buttons: {
    onClick: () => void;
    title: string;
    type: BUTTONS_LOCK;
  }[];
  isFetchLoading: boolean;
  notification: {
    description: string;
    title: string;
    type: NotificationType;
  };
};

export const MSAResultScreen = ({ buttons, isFetchLoading, notification: { description, title, type } }: Props) => {
  const { isMobile } = useBreakpointState();

  return isFetchLoading ? (
    <ResultScreenSkeleton />
  ) : (
    <>
      <AdaptiveOperationStatus description={description} title={title} type={type} />
      <Flex flexDirection={isMobile ? 'column' : 'row'} maxWidth={isMobile ? '100%' : FIELD_WIDTH.xl}>
        {buttons.map(({ onClick, title, type }, index) => {
          const indent = index > 0 ? INDENT.m : undefined;
          const ButtonComponent = type === BUTTONS_LOCK.SECONDARY ? ButtonOutlined : ButtonContained.Loader;

          return (
            <Box ml={isMobile ? undefined : indent} mt={isMobile ? indent : undefined} key={`${title}_${type}`}>
              <ButtonComponent data-test-id={getDataTestId(type, 'blockCardButton')} onClick={onClick}>
                {title}
              </ButtonComponent>
            </Box>
          );
        })}
      </Flex>
    </>
  );
};
