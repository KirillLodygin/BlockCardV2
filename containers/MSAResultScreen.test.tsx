import { shallow } from 'enzyme';
import React from 'react';

import { ButtonContained, ButtonOutlined } from '@vtb/ui-kit/adaptive/atoms/button';
import { NOTIFICATION_TYPE } from '@vtb/ui-kit/adaptive/molecules/notification';

import { BUTTONS_LOCK } from '../types';
import { MSAResultScreen } from './MSAResultScreen';

const mockOnClose = jest.fn();
const mockOnBack = jest.fn();

const defaultProps = {
  buttons: [
    {
      onClick: mockOnClose,
      title: 'asd',
      type: BUTTONS_LOCK.CONTINUE,
    },
    {
      onClick: mockOnBack,
      title: 'qwe',
      type: BUTTONS_LOCK.SECONDARY,
    },
  ],
  isFetchLoading: false,
  notification: {
    description: 'qwe',
    title: 'zxc',
    type: NOTIFICATION_TYPE.info,
  },
};

describe('MSAResultScreen', () => {
  beforeEach(() => {
    mockOnClose.mockReset();
    mockOnBack.mockReset();
  });

  it('renders with all props', () => {
    const component = shallow(<MSAResultScreen {...defaultProps} />);

    expect(component).toMatchSnapshot();
  });

  it('renders with isFetchLoading=true', () => {
    const component = shallow(<MSAResultScreen {...defaultProps} isFetchLoading />);

    expect(component).toMatchSnapshot();
  });

  it('calls onClose', () => {
    const component = shallow(<MSAResultScreen {...defaultProps} />);

    component.find(ButtonOutlined).simulate('click');

    expect(mockOnBack).toHaveBeenCalledTimes(1);
  });

  it('calls onCanfirm', () => {
    const component = shallow(<MSAResultScreen {...defaultProps} />);

    component.find(ButtonContained.Loader).simulate('click');

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });
});
