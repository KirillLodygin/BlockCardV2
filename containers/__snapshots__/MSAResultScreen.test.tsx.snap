// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MSAResultScreen renders with all props 1`] = `
<Fragment>
  <AdaptiveOperationStatus
    description="qwe"
    title="zxc"
    type="info"
  />
  <Styled(styled.div)
    flexDirection="row"
    maxWidth="544px"
  >
    <styled.div
      key="asd_continue"
    >
      <ButtonContainedLoader
        data-test-id="continue_blockcardbutton"
        onClick={[MockFunction]}
      >
        asd
      </ButtonContainedLoader>
    </styled.div>
    <styled.div
      key="qwe_secondary"
      ml="32px"
    >
      <ButtonOutlined
        data-test-id="secondary_blockcardbutton"
        onClick={[MockFunction]}
      >
        qwe
      </ButtonOutlined>
    </styled.div>
  </Styled(styled.div)>
</Fragment>
`;

exports[`MSAResultScreen renders with isFetchLoading=true 1`] = `<ResultScreenSkeleton />`;
